src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -50,6 +51,39 @@ public Task Write100ContinueAsync(int streamId)
             return Task.CompletedTask;
         }
 
+        public void WriteHeaders(int streamId, int statusCode, IHeaderDictionary headers)
+        {
+            lock (_writeLock)
+            {
+                _outgoingFrame.PrepareHeaders(Http2HeadersFrameFlags.NONE, streamId);
+
+                var done = _hpackEncoder.BeginEncode(statusCode, EnumerateHeaders(headers), _outgoingFrame.Payload, out var payloadLength);
+                _outgoingFrame.Length = payloadLength;
+
+                if (done)
+                {
+                    _outgoingFrame.HeadersFlags = Http2HeadersFrameFlags.END_HEADERS;
+                }
+
+                Write(_outgoingFrame.Raw);
```

Comment:
Can this Write block on IO? It better not. Flushing headers never blocks for HTTP/1.1.

Reply:
It's not different than the sync `Write` in `OutputProducer`, except there's no callback involved.

Reply:
So it more like append?

Reply:
Yeah. From looking at it Write doesn't called flush so there's no back pressure. I would probably rename Write to Append in this method to make it clear it's not doing any IO.

I don't see who's calling WriteHeadersAsync which does flush. I would probably remove that method. Today, we never handle IO errors or back pressure when writing headers. If anything like that happens we let the application deal with it when they flush the first write to the body.